vector/src/Data/Vector/Unboxed/Base.hs

```diff
@@ -253,6 +254,86 @@ instance P.Prim a => G.Vector Vector (UnboxViaPrim a) where
   basicUnsafeCopy (MV_UnboxViaPrim mv) (V_UnboxViaPrim v) = G.basicUnsafeCopy mv v
   elemseq _ = seq
 
+-- | Isomorphism between type @a@ and its representation in unboxed
+-- vector @b@.
+class Isomorphic a b where
```

Comment:
Considering that this is designed specifically for `Unbox` class, I can suggest two things:

* Add `Unbox` as a constraint for `b`: 
```haskell
class Unbox b => Isomorphic a b where
```
* Change name to something like `UnboxIso` or `IsoUnbox`

Reply:
Is it right to make this class only for unboxed vectors though? What if someone wants partially unbox ```data Foo = Foo Int Int Double Text``` as some sort of a hybrid vector of three unboxed vectors and one boxed?

Reply:
As long as it's used with `As` newtype `Unbox` constraint on `b` is necessary: 
```haskell
instance (Isomorphic a b, Unbox b) => M.MVector MVector (As a b) where
instance (Isomorphic a b, Unbox b) => G.Vector Vector (As a b) where
```
It's of course possible to add more newtypes. For example one for sneaking boxed representations:

```haskell
import qualified Data.Vector as V
import qualified Data.Vector.Mutable as MV

newtype Boxed a = Boxed a
newtype instance Vector (Boxed a) = V_Box (V.Vector a)
newtype instance MVector s (Boxed a) = MV_Box (MV.Vector s a)
{- necessary instances -}

data Foo = Foo Int Text

instance VU.Isomorphic Foo (Int,Boxed Text) where
   toURepr (Foo i s) = (i,Boxed a)
   fromURepr (i,Boxed a) = Foo i (Boxed a)
{- rest of boilerplate -}
```

Reply:
I tried adding superclass constraint `class Unbox b => IsoUnbox a b where` but decided against it. It requires adding `Unbox` constraint when defining `IsoUnbox` instances for polymorphic types:

```haskell
instance Unbox a => VU.IsoUnbox (Foo a) (Int,a) where
-- vs
instance VU.IsoUnbox (Foo a) (Int,a) where
```

It only creates busywork for users and `Vector`/`MVector` instances need Unbox constraint either way